### ISSUE TITLE
[no-title]

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9537]
[no-title]<br>
<br>
https://ift.tt/vTdeXE0<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<p><br></p>
<p><a href="https://cs.stanford.edu/people/karpathy/">Andrew Karpathy</a> 最近更新了一篇博文：<a href="http://karpathy.github.io/2019/04/25/recipe/">A Recipe for Training Neural Networks </a>，阅读过后有不少的收获，推荐阅读。本文整理了原博文中比较有用的点，记录在这里。</p>
<p><br></p>
<h3>目录</h3>
<ol>
<li><a href="https://borgwang.github.io/dl/2019/05/01/training-neural-networks.html#1-%E5%88%86%E6%9E%90%E6%95%B0%E6%8D%AE">分析数据</a></li>
<li><a href="https://borgwang.github.io/dl/2019/05/01/training-neural-networks.html#2-%E6%90%AD%E5%BB%BA%E4%B8%80%E4%B8%AA%E5%AE%8C%E6%95%B4%E7%9A%84%E8%AE%AD%E7%BB%83%E9%AA%8C%E8%AF%81-pipeline%E5%BE%97%E5%88%B0%E4%B8%80%E4%B8%AA-baseilne-%E6%A8%A1%E5%9E%8B">搭建一个完整的训练/验证 pipeline，得到一个 baseilne 模型</a></li>
<li><a href="https://borgwang.github.io/dl/2019/05/01/training-neural-networks.html#3-%E4%BA%A7%E7%94%9F%E4%B8%80%E4%B8%AA%E6%9B%B4%E4%BC%98%E7%9A%84%E6%A8%A1%E5%9E%8B">产生一个更优的模型</a></li>
<li><a href="https://borgwang.github.io/dl/2019/05/01/training-neural-networks.html#4-%E5%8E%8B%E6%A6%A8%E6%9C%80%E5%90%8E%E7%9A%84%E9%AA%8C%E8%AF%81%E9%9B%86%E6%80%A7%E8%83%BD%E6%8F%90%E5%8D%87">压榨最后的（验证集）性能提升</a></li>
<li><a href="https://borgwang.github.io/dl/2019/05/01/training-neural-networks.html#5-%E5%85%B6%E4%BB%96%E5%BB%BA%E8%AE%AE">其他建议</a></li>
</ol>
<hr>
<h3>1. 分析数据</h3>
<p><br></p>
<h3>2. 搭建一个完整的训练/验证 pipeline，得到 baseilne 模型</h3>
<p>在上一步对数据有了一定的感觉之后可以开始搭建训练/验证的 pipeline，构建这个 pipeline 的时候要确保万无一失，仔细实现以及检查。选定好验证集和评估方法并在后续的实验中保持固定不变。在此基础上可以使用简单的、经典的算法构建 baseline 模型。</p>
<p>这一步中有以下几点需要注意的：</p>
<ul>
<li>
<p>固定好随机种子（保证实验可复现）</p>
</li>
<li>
<p>参数的初始化。主要网络最后一层参数的设置，如果回归值期望大概在 50 左右，那么将最后一层的 bias 设置在 50 左右；如果分类的概率大概是 0.1，那么将最后一层的 bias 调整至让网络初始能够输出 0.1 附近的参数。这样做的意义是引入一点人为先验，帮助网络参数初始在一个相对较好的位置，可以加速收敛。</p>
</li>
<li>
<p>设置一个 input-independent baseline。input-independent 是指将数据中的 x 使用 0 或者随机噪声代替，训练得到的一个 baseline 模型。与这个 baseline 模型对比可以看出你当前的模型是否确实从输入数据中提取到了有用的信息</p>
</li>
<li>
<p>设置一个人的 baseline。抽出一部分数据，人肉回归/分类，统计结果作为 baseline。留意在做决策的时候人的关注点在哪</p>
</li>
<li>
<p>overfit 一个 batch。这是一个检查性的操作，通过加大模型的容量，在一小个 batch 的数据上面训练，看能不能将 training loss 能否降到非常接近 0。若不能则说明代码流程中可能存在某些 bug 需要仔细检查</p>
</li>
<li>
<p>验证数据的准确性的时候，从数据进入模型的入口进行检查（如 sklearn API 中的 <code>model.fit(x, y)</code>）</p>
</li>
<li>
<p>使用一个<strong>固定的具有代表性的</strong>测试集，通过不同的模型在测试集上的效果可以给你一些 intuition，哪些模型是 work 的，哪些模型不 work</p>
</li>
</ul>
<p><br></p>
<h3>3. 产生一个更优的模型</h3>
<p>上个阶段我们已经对数据集有了一定的理解，有了一个完整的训练/验证 pipeline，给定一个模型可以产生一个可靠的指标体现模型的性能。我们有了几个 baseline 模型，一个是 input-independent 的模型，一些非常简单的 baseline，可能还有人工的 baseline。因此这个阶段就是在上个阶段的基础上去产生一个更优的模型。</p>
<p>这个阶段可以分成两步：</p>
<ol>
<li>focus on training loss，先专注于降低训练集误差
<ul>
<li>模型选型：在项目的前期不要使用太 fancy 的模型，优先选择经典的被验证过效果好的模型</li>
<li><u>如果同时多个不同的特征/信号/技巧增加到模型中，不要一股脑全部放进去，尽量 one by one，并且同时记录每个操作带来的影响，保留那些给模型性能带来提升的操作</u></li>
</ul>
</li>
<li>focus on validation loss，<strong>牺牲一部分训练误差来换取验证集误差</strong>
<ul>
<li>获取更多的数据（第一选择）、在没办法获得更多数据的情况下考虑数据增强</li>
<li>Pre-train 几乎没有坏处</li>
<li>更小的 batch size</li>
<li>dropout/early-stopping</li>
<li>一个大的 early-stopping 的模型通常要比一个训练更久的小模型泛化性能更好</li>
</ul>
</li>
</ol>
<p><br></p>
<h3>4. 压榨最后的（验证集）性能提升</h3>
<p><br></p>
<h3>5. 其他建议</h3>
<ul>
<li>在实际的项目中，不需要一开始就将代码写的十分通用（通常情况下你也做不到），而是先写一个针对具体问题的非常 specific 的，正确的 run 起来，后面再考虑泛化的事情。</li>
</ul>
<p><br><br></p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live. <br>
<br>
June 16, 2023 at 03:57PM
